File: dsa/7.linked_list/707_Design_Linked_List.java
class Node {
    public int val;
    public Node next;

    public Node(int val){
        this.val = val;
        this.next = null;
    }
}
class MyLinkedList {
    Node head;
    int len;

    public MyLinkedList() {
        this.head = null;
        this.len = 0;
    }
    
    public int get(int index) {
        if (index >= this.len){
            return -1;
        }

        Node indNode = this.head;
        for (int i = 0; i < index; i++){
            indNode = indNode.next;
        }
        return indNode.val;
    }
    
    public void addAtHead(int val) {
        Node newNode = new Node(val);
        newNode.next = this.head;
        this.head = newNode;
        this.len += 1;
    }
    
    public void addAtTail(int val) {
        if (this.head == null){
            this.addAtHead(val);
        }
        else{
            Node newNode = new Node(val);
            Node indNode = this.head;
            while (indNode.next != null){
                indNode = indNode.next;
            }
            indNode.next = newNode;
            this.len += 1;
        }
        
    }
    
    public void addAtIndex(int index, int val) {
        if (index == 0){
            this.addAtHead(val);
            return;
        }

        if (index == this.len){
            this.addAtTail(val);
            return;
        }

        if (index > this.len){
            return;
        }

        Node newNode = new Node(val);
        Node indNode = this.head;

        for (int i = 0; i < index - 1; i++){
            indNode = indNode.next;
        }

        Node nextNode = indNode.next;
        indNode.next = newNode;
        newNode.next = nextNode;
        this.len += 1;
    }
    
    public void deleteAtIndex(int index) {
        if (index >= this.len){
            return;
        }

        if (index == 0){
            this.head = this.head.next;
            this.len -= 1;
            return;
        }

        Node indNode = this.head;
        for (int i = 0; i < index - 1; i++){
            indNode = indNode.next;
        }
        if (index == this.len - 1){
            indNode.next = null;
        }
        else {
            Node newNextNode = indNode.next.next;
            indNode.next = newNextNode;
        }
        this.len -= 1;
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */

File: dsa/7+8.linked_list/707_Design_Linked_List.py
class Node:
    def __init__(self, val: int):
        self.val: int = val
        self.next: Node = None
        self.prev: Node = None
    
class MyLinkedList:

    def __init__(self):
        self.head: Node = None
        self.tail: Node = None

    def get(self, index: int) -> int:
        pointer_node: Node = self.head
        for _ in range(0, index, 1):
            if not pointer_node:
                return -1
            pointer_node = pointer_node.next
        
        return pointer_node.val if pointer_node else -1


    def addAtHead(self, val: int) -> None:
        new_node: Node = Node(val)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.head.prev = new_node
            new_node.next = self.head
            self.head = new_node


    def addAtTail(self, val: int) -> None:
        new_node: Node = Node(val)
        if not self.tail:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node


    def addAtIndex(self, index: int, val: int) -> None:
        pointer_node: Node = self.head

        for ind in range(0, index, 1):
            if not pointer_node and ind < index:
                return
            pointer_node = pointer_node.next
        
        if not pointer_node:
            self.addAtTail(val)
            return
        
        prev_node: Node = pointer_node.prev
        if not prev_node:
            self.addAtHead(val)
            return
        
        new_node: Node = Node(val)
        prev_node.next = new_node
        new_node.prev = prev_node

        new_node.next = pointer_node
        pointer_node.prev = new_node


    def deleteAtIndex(self, index: int) -> None:
        pointer_node: Node = self.head

        for ind in range(0, index, 1):
            if not pointer_node and ind < index:
                return
            pointer_node = pointer_node.next
        
        if not pointer_node:
            return 
        
        if not pointer_node.prev:
            self.head = pointer_node.next
            if self.head: 
                self.head.prev = None
            pointer_node.next = None
            return
        
        if not pointer_node.next:
            self.tail = pointer_node.prev
            self.tail.next = None
            pointer_node.prev = None
            return
        
        next_node: Node = pointer_node.next
        prev_node: Node = pointer_node.prev

        prev_node.next = next_node
        next_node.prev = prev_node

        pointer_node.next = None
        pointer_node.prev = None


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)